test-d/destructurable.ts

```diff
@@ -0,0 +1,50 @@
+import {expectType} from 'tsd';
+import {type Destructurable} from '../index';
+
+// Typological use cases
+type Success = {
+	type: 'success';
+	value: string;
+};
+type Failure = {
+	type: 'failure';
+	error: Error;
+};
+type Result = Destructurable<Success | Failure>;
+function doSomething(): Result {
+	return Math.random() > 0.5
+		? {type: 'success', value: 'foo'}
+		: {type: 'failure', error: new Error('Something wrong')};
+}
+
+// The type of `result` should be the union of the destructured types
+const {type, value, error} = doSomething();
+expectType<'success' | 'failure'>(type);
+expectType<string | undefined>(value);
+expectType<Error | undefined>(error);
+
+// Narrowing should work
+if (type === 'success') {
+	expectType<string>(value);
+	expectType<undefined>(error);
+} else {
+	expectType<Error>(error);
+	expectType<undefined>(value);
+}
+
+// Readonly properties should be preserved
+declare const result1: Destructurable<{readonly x: number}>;
+expectType<Expect1>(result1);
+type Expect1 = {readonly x: number};
+
+// If both writable properties and readonly properties are present, it should be converted to writable properties
```

Comment:
Sure about that? My intuitive answer would be: If `foo` is `readonly` in one type, it should also be `readonly` in the union. This seems like a more secure default to me. Otherwise you would be able to accidentally ignore the `readonly`.

Reply:
Thank you for your review and for bringing up this point. I apologize if the comment gave you a wrong impression. Please allow me to clarify.

In this situation, the properties that will be converted to writable are only those that are _complemented to the objects that do not originally have the property_. The properties that originally have a `readonly` modifier will remain `readonly`.

To provide more clarity, when properties are added to objects that originally do not have those properties, they become optional with the `undefined` type. However, no `readonly` modifier is added to these properties. This is in accordance with the TypeScript compiler's type inference methods. Consider the following example:

```typescript
function get_result(n: number) {
  if (n === 1) return { type: 1 as const, foo: 123 };
  else if (n === 2)
    return { type: 2, foo: 123 } as { type: 2; readonly foo: number };
  else return { type: 3 as const, bar: 123 };
}

type Result = ReturnType<typeof get_result>;
```

The TypeScript compiler will infer the `Result` type as:

```typescript
{
    type: 2;
    readonly foo: number;
} | {
    type: 1;
    foo: number;
    bar?: undefined;
} | {
    type: 3;
    bar: number;
    foo?: undefined;
}
```

As you can see, in the first part of the union, the `foo` property remains `readonly`, while the other two parts have `foo` as an optional undefined property, but without the `readonly` modifier. This ensures that any property that was originally `readonly` will still be treated as such.